File: naked-generator/Primary/py/steer.py
from enum import Enum
from struct import pack, unpack
from collections import namedtuple


class Tlm_Status(Enum):
	ON = 0
	OFF = 1


class Race_Type(Enum):
	ACCELERATION = 0
	SKIDPAD = 1
	AUTOCROSS = 2
	ENDURANCE = 3


class Car_Status(Enum):
	IDLE = 0
	SETUP = 1
	RUN = 2


class Inverter_Status: Inverter_L(Enum):
	OFF = 0
	IDLE = 1
	ON = 2


class Inverter_Status: Inverter_R(Enum):
	OFF = 0
	IDLE = 1
	ON = 2


TLM_STATUS = namedtuple('TLM_STATUS', 'tlm_status race_type driver circuit')
TLM_STATUS_schema = '<ccBB'

CAR_STATUS = namedtuple('CAR_STATUS', 'car_status Inverter_status: inverter_l Inverter_status: inverter_r')
CAR_STATUS_schema = '<ccc'


def serialize_TLM_STATUS(struct: TLM_STATUS) -> bytearray:
	return pack(TLM_STATUS, *tuple(struct))


def serialize_CAR_STATUS(struct: CAR_STATUS) -> bytearray:
	return pack(CAR_STATUS, *tuple(struct))


def deserialize_TLM_STATUS(buffer: bytearray) -> TLM_STATUS:
	return TLM_STATUS._make(unpack(TLM_STATUS_schema, buffer))


def deserialize_CAR_STATUS(buffer: bytearray) -> CAR_STATUS:
	return CAR_STATUS._make(unpack(CAR_STATUS_schema, buffer))

File: naked-generator/Primary/c/steer.c
#include <string.h>
#include <assert.h>
#include <stdio.h>

#include "steer.h"

void serialize_TLM_STATUS(TLM_STATUS* tlm_status, uint8_t* buffer, size_t buf_len) {
	assert(buf_len >= sizeof(TLM_STATUS));
	memcpy(buffer, tlm_status, sizeof(STEER_STATUS));
}
void serialize_CAR_STATUS(CAR_STATUS* car_status, uint8_t* buffer, size_t buf_len) {
	assert(buf_len >= sizeof(CAR_STATUS));
	memcpy(buffer, car_status, sizeof(STEER_STATUS));
}

void deserialize_TLM_STATUS(uint8_t* buffer, size_t buf_len, TLM_STATUS* tlm_status) {
	assert(buf_len >= sizeof(TLM_STATUS));
	memcpy(tlm_status, buffer, sizeof(STEER_STATUS));
}
void deserialize_CAR_STATUS(uint8_t* buffer, size_t buf_len, CAR_STATUS* car_status) {
	assert(buf_len >= sizeof(CAR_STATUS));
	memcpy(car_status, buffer, sizeof(STEER_STATUS));
}
